src/commissioner/Events.h

```diff
@@ -0,0 +1,316 @@
+    CHIP_ERROR Set(const ByteSpan buffer)
+    {
+        if (mCapacity < buffer.size())
+        {
+            ReturnErrorOnFailure(Allocate(buffer.size()));
+        }
+        memmove(mData.get(), buffer.data(), buffer.size());
+        mLen = buffer.size();
+        return CHIP_NO_ERROR;
+    }
+    void Clear()
+    {
+        mData     = std::shared_ptr<uint8_t>(nullptr);
+        mCapacity = 0;
+        mLen      = 0;
+    }
+    CHIP_ERROR Allocate(size_t size)
+    {
+        mData = std::shared_ptr<uint8_t>(static_cast<uint8_t *>(chip::Platform::MemoryAlloc(size)), chip::Platform::MemoryFree);
+        if (mData.get() == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        mLen      = 0;
+        mCapacity = size;
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    Platform::SharedPtr<uint8_t> mData;
+    size_t mCapacity = 0;
+    size_t mLen      = 0;
+};
+
+struct SharedString
+{
+    const char * Get() const { return mData.get(); }
+    CHIP_ERROR Set(const char * cStr)
+    {
+        if (cStr == nullptr)
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        const size_t len = sizeof(*cStr) * (strlen(cStr) + 1);
+        if (mCapacity < len)
+        {
+            ReturnErrorOnFailure(Allocate(len));
+        }
+        memmove(mData.get(), cStr, len);
+        mLen = len;
+        return CHIP_NO_ERROR;
+    }
+    void Clear()
+    {
+        mData     = std::shared_ptr<char>(nullptr);
+        mCapacity = 0;
+        mLen      = 0;
+    }
+    CHIP_ERROR Allocate(size_t len)
+    {
+        mData = std::shared_ptr<char>(static_cast<char *>(chip::Platform::MemoryAlloc(len)), chip::Platform::MemoryFree);
+        if (mData.get() == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        mLen      = 0;
+        mCapacity = len;
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    Platform::SharedPtr<char> mData;
+    size_t mCapacity = 0;
+    size_t mLen      = 0;
+};
+
+template <size_t MaxSize>
+struct CapacityBoundSharedBuffer : public SharedBuffer
+{
+    CHIP_ERROR Allocate(size_t size)
+    {
+        if (size > MaxSize)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return SharedBuffer::Allocate(size);
+    }
+    CHIP_ERROR Set(const ByteSpan buffer)
+    {
+        if (buffer.size() > MaxSize)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return SharedBuffer::Set(buffer);
+    }
+};
+
+template <size_t MaxSize>
+struct CapacityBoundStaticBuffer
+{
+    const ByteSpan Get() const { return ByteSpan(mBuffer, mBuffer.Length()); }
+    void Clear() { mBuffer.SetLength(0); };
+    CHIP_ERROR Set(ByteSpan buffer)
+    {
+        ReturnErrorOnFailure(mBuffer.SetLength(buffer.size()));
+        memmove(mBuffer, buffer.data(), buffer.size());
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    Crypto::CapacityBoundBuffer<MaxSize> mBuffer;
+};
+
+// Events
+struct Success
```

Comment:
Is there a way to make these generic events more greppable?

I find the overloading of `Event` / `Context` symbols by the StateMachine templates and implementations makes the code harder to follow than it could be.

Reply:
Please see https://github.com/project-chip/connectedhomeip/pull/13294/commits/ef37720e6ad49f100e3fbccbf65608ab7c191589, which should make this much clearer.